api/rosetta/validator.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-playground/validator/v10"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+)
+
+var (
+	// Special case in the validator library that shouldn't really happen with correct usage.
+	errInvalidValidation = errors.New("invalid validation input")
```

Comment:
Can you expand on this? I don't understand why we need a sentinel for it if it can't happen. What do you mean exactly by `shouldn't happen`? That it can only happen if we misuse the component in the code, I assume?

Reply:
Yes, it only happens if you call `validate.Struct(x)` where x is a non-struct.

This won't happen in our code, but the validation library defines two classes of errors - `ValidationErrors` (actual errors) and `InvalidValidationError` (invalid input).
It felt wrong to ignore it, even if it will not happend.

Still, since it's an equivalent of an internal error for "unable to validate X", I didn't think there's value in relaying the exact error to the API user.

I'm not really happy with either options TBH, it's really nagging me.

Reply:
If only we had generics and we could have a compile error when giving a non-struct to `validate.Struct`, wouldn't that be something? :p

In the meantime, I guess we can keep this. Hopefully in 1 year or so, we can remove this error :p